src/mpid/ch4/netmod/ofi/ofi_rma.h

```diff
@@ -384,14 +384,114 @@ static inline int MPIDI_OFI_do_put(const void *origin_addr,
     }
 
     /* noncontiguous messages */
-    if (sigreq)
-        mpi_errno =
-            MPIDIG_mpi_rput(origin_addr, origin_count, origin_datatype, target_rank, target_disp,
-                            target_count, target_datatype, win, sigreq);
-    else
-        mpi_errno =
-            MPIDIG_mpi_put(origin_addr, origin_count, origin_datatype, target_rank, target_disp,
-                           target_count, target_datatype, win);
+    MPI_Aint origin_density, target_density;
+    MPIDI_Datatype_get_density(origin_datatype, origin_density);
+    MPIDI_Datatype_get_density(target_datatype, target_density);
+
+    /* Sparse targets should unpack from temporary buffer. Let the generic AM code handle it. */
+    if (target_density < MPIR_CVAR_CH4_OFI_IOVEC_MIN_SEG_SIZE)
+        goto am_fallback;
+
+    /* target is high density, create iovecs representing target */
+    struct iovec *target_iov;
+    MPI_Aint num_iov, actual_iov_bytes;
+    int target_iov_len;
+    MPIR_Typerep_iov_len(target_count, target_datatype, target_bytes, &num_iov);
+    target_iov = MPL_malloc(sizeof(struct iovec) * num_iov, MPL_MEM_RMA);
+    offset = target_disp * MPIDI_OFI_winfo_disp_unit(win, target_rank);
+    MPIR_Typerep_to_iov((void *) (MPIDI_OFI_winfo_base(win, target_rank) + offset), target_count,
+                        target_datatype, 0, target_iov, num_iov, target_bytes, &target_iov_len,
+                        &actual_iov_bytes);
+
+    /* allocate request */
+    MPIDI_OFI_win_request_t *req = MPIDI_OFI_win_request_create();
+    MPIR_ERR_CHKANDSTMT((req) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
+    req->event_id = MPIDI_OFI_EVENT_ABORT;
+    req->next = MPIDI_OFI_WIN(win).syncQ;
+    MPIDI_OFI_WIN(win).syncQ = req;
+
+    /* Target is high density while origin is low. Pack origin data into temporary buffer. */
+    if (origin_density < MPIR_CVAR_CH4_OFI_IOVEC_MIN_SEG_SIZE) {
+        MPI_Aint actual_pack_bytes;
+        req->noncontig.pack_buffer = MPL_malloc(origin_bytes, MPL_MEM_BUFFER);
+        MPIR_Typerep_pack(origin_addr, origin_count, origin_datatype, 0, req->noncontig.pack_buffer,
+                          origin_bytes, &actual_pack_bytes);
```

Comment:
Do we need to limit the temporary buffer size here?  This can be as large as `origin_bytes`.

Reply:
Same issue here. If multiple Typerep calls at OK, should be straightforward change to make.

Reply:
Using multiple Typerep calls here is not a problem.  It's been designed to use it in this manner.

Reply:
So if we want to save memory, we would have to wait for each operation to complete before reloading the pack buffer. Perhaps we just fallback to AM if origin density is low and `origin_bytes > max_msg_size`?

Reply:
Would the AM code need an extra copy?  If yes, that'd be expensive for large messages.

Reply:
Technically no, since ch4/ofi AM implementation uses FI_READ for payload transfer. But I'm going to need to abstract partial data operations anyway, so I might as well build it out for this, too.